Deploy button

tonykploomber opened the pull request:
Continue of https://github.com/ploomber/jupysql-plugin/pull/24

Also in this PR:
- Add Copy to Clickboard function

https://github.com/ploomber/jupysql-plugin/assets/123580782/542aaca5-72cd-4744-b547-22e85062fadd

===

After reviewing: 
If we finalize this is the flow, we can explore how to write the testing for this react component, probably we need to have another github action to run the testing against to frontend component

edublancas reviewed (changes requested):
looks great!

some feedback:
---

when I click on deploy notebook the window takes some time to load. why is that? we should provide feedback to the user so they don't think this is broken

![image](https://github.com/ploomber/jupysql-plugin/assets/989250/8402c0f6-9d13-4a2e-8543-046445f1eb52)

---

currently, the API key and "get deployment URL" are displayed in the same view, but they should be separated. i.e., if there is no API key, we should ask for one, store it, and then proceed to ask for confirmation to deploy the notebook.

but if there's an API key, there's no need to display the API key text field, we can just ask for confirmation to deploy

let me create a diagram with the detailed flow so we can agree on specifics. I'll post it in the issue and tag you

--- 

I see there are a  bunch of new js dependencies added. that's fine but we should ensure that if we choose to go with this mui library, it's gonna be enough for any other UI  needs we have for other all out plugins. looks pretty complete so it looks like a good choice

@yafimvo: you've been working on the connector widget. thoughts? do you think it makes sense to make mui a dependency or should be build the UI components ourselves.

yafimvo commented:
@tonykploomber I agree with @edublancas , it looks great!


> I see there are a bunch of new js dependencies added. that's fine but we should ensure that if we choose to go with this mui library, it's gonna be enough for any other UI needs we have for other all out plugins. looks pretty complete so it looks like a good choice
> 
> @yafimvo: you've been working on the connector widget. thoughts? do you think it makes sense to make mui a dependency or should be build the UI components ourselves.

I agree. Basically [MUI](https://mui.com/material-ui/getting-started/overview/) is a great library, allowing easy implementation of material UI principles by using out-of-the-box components, API, and icons with react (which is also great e.g. hooks, HOC, JSX, etc...).

Using React and this library can definitely save us some time and make sure our JupySQL-Plugin design is consistent.

However, we need to make sure it works for other cases, such:

1. Creating a UI widget we can use in a cell (e.g. `TableWidget()` and `FormWidget()` like [here](https://github.com/ploomber/jupysql-plugin/blob/main/ui-tests/notebooks/widgets.ipynb))
2. Communication with the kernel - we should be able to pass information from back to front, and from front to back (like between [form.ts](https://github.com/ploomber/jupysql-plugin/blob/main/src/widgets/form.ts) and [FormWidget](https://github.com/ploomber/jupysql-plugin/blob/main/jupysql_plugin/_widgets.py#L7)).
3. For the 1st and 2nd bullets - how easy it is? Is there enough documentation for it? Can we still use the [official examples](https://github.com/jupyterlab/extension-examples)? I think these questions are crucial since this is not a regular React component, it's a jupyterlab extension, and if something is not aligned/supported with their architecture it would be very hard to move fast.
4. We need to make sure it works with a classic notebook as well. Would it require writing any additional code?
5. Test units - does it work with the existing testing method or should we implement something new?

There are also some design topics we would probably need to discuss (folder structure, re-use of code, etc...) but I think it's less important for now.

edublancas commented:
@yafimvo: fair point. I have little experience with react so I cannot make a call with the information we have.

@tonykploomber: can you investigate the points that @yafimvo raised?

- can we use it in a widget? e.g., TableWidget
- can we pass information to the kernel?
- how hard this is?
- will this work in the classic notebook? 
- can we use the existing ui-tests/ mechanism?

yafimvo commented on src/dialog.tsx:
```diff
@@ -0,0 +1,137 @@
+                setAPIValidStatus("success")
+            }
+            setIsLoadingRemoteAPI(false);
+            clearInterval(simulateAPICall)
```
Why do we use interval? can we use setTimeout instead?

yafimvo reviewed (changes requested):
UI/UX
1. The "Deploy Notebook" Button may be easily missed since it blends in with all the other buttons (e.i. "Format SQL"). It's a core feature and I think we can somehow make it more "noticeable" (maybe replace between "Deoploy Notebook" and "Format SQL"?)
2. There is a small delay after clicking the "Deploy Notebook" with this UI
![image](https://github.com/ploomber/jupysql-plugin/assets/103026689/9647ddbf-e6f3-4385-bc92-be97e393a299)
Since we are using MUI, we can add a [circular progress](https://mui.com/material-ui/react-progress/#circular)
3. This UI is a bit confusing to me since I'm not sure what to do.
![image](https://github.com/ploomber/jupysql-plugin/assets/103026689/ac2beaf3-19b0-455c-aa99-c981000654a9)

If the user didn't set the API key - I think it makes sense to show this UI but I think we need to disable the "GET DEPLOYMENT URL" button (and "OK" which I'm not sure what it does) and add some instructions above this field and maybe explain how to get this API key.

If the user did set the API key - Why do we need to show the API key?

I think we shouldn't allow the user to edit this field so easily if the API key was set,

Suggestion:
We can disable this field and add an "edit" button. If the user clicks edit - we allow to edit this field and we replace the "edit" button with 2 buttons: "cancel" and "save".

We can also add a small mark indicating that the API was set.

Example (draft):
![image](https://github.com/ploomber/jupysql-plugin/assets/103026689/3ab38ed7-6ecd-45f9-bc4e-0c5c55071ec5)

4. "GET DEPLOYMENT URL" - At this point, the user doesn't know where the notebook would be deployed so I think the action is not clear enough. Is clicking on this button would deploy the notebook? what's this URL for? 

![image](https://github.com/ploomber/jupysql-plugin/assets/103026689/38af0856-a71e-4d23-9b35-7ae199a837e8)

Suggestion: 

* Add some instructions
* Change the action name to something like: "Deploy notebook" 
* After the user clicks "deploy notebook" - we can make a redirect button to the URL (maybe we should call it "Dashboard"). Do we actually need to allow the user to copy the URL?
* If we keep the URL input - I would also change its design to match the API Key input to make the design consistent.

tonykploomber commented:
Updated TODO List:
1. Update the UI with [new flow](https://github.com/ploomber/jupysql-plugin/issues/17#issuecomment-1583444146) 
2. Investigate how to acces Notebook meta in jupyter extension
3. Investigate how to communicate with ploomber-core to store API key

tonykploomber commented:
@edublancas 
I have some progress on [access Notebook meta in jupyter extension](https://github.com/ploomber/jupysql-plugin/issues/34)

tonykploomber commented:
@yafimvo 
I am still working on the research part on other topic.
But I can give you simple conclusion to the question:
>Creating a UI widget we can use in a cell (e.g. TableWidget() and FormWidget() like [here](https://github.com/ploomber/jupysql-plugin/blob/main/ui-tests/notebooks/widgets.ipynb))

React Solution:
Someone provide the [reaction](https://github.com/widgetti/reacton) to let react be wrapped as UI component in the jupyter cell, but I don't think this is the best approach since all the development work will rely on this library, we are not sure what's the limitation before developing extensively about this

Python Solution:
There is someone provide the [itables](https://github.com/mwouts/itables) to make the responsive table (pagination). I think probably we can integrate this into our library, seems it's more native solution?

edublancas commented:
when mentioning widgets, I think @yafimvo mean if we could MUI for widgets as well. i.e., use mui to design widgets, I don't think it'll be an issue since I've used other libraries in the past for creating widgets

tonykploomber commented:
Some conclusion @yafimvo and I discussed:

---

Implement widgets with React, there are two scopes we were discussing:
1. jupyter extension with React, this is supported by [JupyterLab](https://jupyterlab.readthedocs.io/en/latest/extension/virtualdom.html) and it's doable, we are trying to implement the unit test by jest and playwright e2e testing to this PR
2. ipywidget with React, this is probably not an easy approach. The closest one is using [reaction](https://github.com/widgetti/reacton), current existing solutions are still using vanilla javascript HTML to implement this part.
  However, we might still use [Material Design](https://m3.material.io/) to enrich the CSS/HTML part. I can research this after this is done

---
Will report any new progress here

yafimvo commented:
> 1. jupyter extension with React, this is supported by [JupyterLab](https://jupyterlab.readthedocs.io/en/latest/extension/virtualdom.html) and it's doable, we are trying to implement the unit test by jest and playwright e2e testing to this PR

@tonykploomber 
Is it possible to use only playwright? Why do we need both?

tonykploomber commented:
@edublancas Please check this flow, I am going to implement testing part now

@yafimvo jest is for unit testing, playwright is for e2e testing. I think they have different purpose

===

New Flow:
1. No API Key + Not First Time deployment

https://github.com/ploomber/jupysql-plugin/assets/123580782/71de4d71-bd47-431d-b8d7-58cab47345c0

2. No API Key + First Time deployment

https://github.com/ploomber/jupysql-plugin/assets/123580782/2cb9cfa6-1839-4cdb-b25b-46df49696a03

3. Invalid API Key

https://github.com/ploomber/jupysql-plugin/assets/123580782/bf586248-1ff3-4821-b087-b95c85e3e174

4. API Key + First Time deployment

https://github.com/ploomber/jupysql-plugin/assets/123580782/e180ed86-ab54-4403-996c-08ddb8c49ce0

5. API Key + Not First Time deployment

https://github.com/ploomber/jupysql-plugin/assets/123580782/9b6b8ab4-4f07-47f0-8687-efa4d34f1082

edublancas commented:
> @yafimvo jest is for unit testing, playwright is for e2e testing. I think they have different purpose

yeah, we have to use both

---

@tonykploomber feedback on the videos you shared

> 1. No API Key + Not First Time deployment

I see a black tooltip when typing the API key, let's remove it. Instead of showing `URL: {URL}`, let's put `Check your deployment status here: {URL}` <- note that these recommendations apply to other scenarios as well.

> 2. No API Key + First Time deployment

Since this is first time deployment, let's show a message: `Clicking on deploy will upload your notebook to Ploomber Cloud servers.` and the blue button for deployment should appear below.

For cases where it's not the first deployment, we can skip the screen altogether and just skip to this:

![image](https://github.com/ploomber/jupysql-plugin/assets/989250/5591cea7-4335-4f9a-95f1-08c5a4b54e91)

we can get rid of this screen:

![image](https://github.com/ploomber/jupysql-plugin/assets/989250/00991ff2-4d6f-43fc-aee2-2fd770608a9a)


> 3. Invalid API Key

Change: New API Key -> `Click here to get an API Key`

tonykploomber commented:
@edublancas 
How do you test the playwright locally, did you just execute `nox --session test --no-install --reuse-existing-virtualenvs` on local?

edublancas commented:
yes, you can use nox. with the arguments you showed, it'll use the existing env. if you omit some of those, you can tell nox to create a new environment.

or you can run them directly: https://github.com/ploomber/jupysql-plugin/tree/main/ui-tests#run-the-tests

tonykploomber commented:
Flow Update:

1. No API Key + Not First Time deployment
Update: After entering API Key, it should deploy directly

https://github.com/ploomber/jupysql-plugin/assets/123580782/73f9cec2-62d3-4c79-9c68-c32d8f90b472

2. No API Key + First Time deployment

https://github.com/ploomber/jupysql-plugin/assets/123580782/1730b8f2-2280-4cb1-87b4-9e306686c48e

4. API Key + First Time deployment

This will deploy directly, no confirm box anymore

https://github.com/ploomber/jupysql-plugin/assets/123580782/b96e2a50-4a36-4806-b771-d691d79cdad1

===

A. For the black tooltip, do you mean this autocomplete box? If yes, this was from browser
<img width="367" alt="Screenshot 2023-06-26 at 4 49 42 PM" src="https://github.com/ploomber/jupysql-plugin/assets/123580782/faae2844-76fb-40b1-9e1b-c4d2d30248eb">

B. `Check your deployment status here: {URL}`, for this I have to separate two lines, like this:
<img width="359" alt="Screenshot 2023-06-26 at 4 51 15 PM" src="https://github.com/ploomber/jupysql-plugin/assets/123580782/82bee631-bf4d-4ccb-9d1b-f9e0b7b11cf0">

tonykploomber commented:
After this button is done, we should hide this button for now since we are still building the backend service?

edublancas commented:
> After this button is done, we should hide this button for now since we are still building the backend service?

the deploy notebook button? 

No. no need to hide it. we'll release this button at the same time as the backend services

tonykploomber commented:
@edublancas 

Addressed some feedback:
1. metadata to store project id:
Looks like we can only store the `project_id` under metadata, tried to store under `metadata->dashboard->project_id` but failed
```
 "metadata": {
  "project_id": "summer-water-6547",
  "kernelspec": {
   ...
  },
  "language_info": {
    ...
   },
  }
 },
```

2. I removed to button in the dialog, I think it's clear since previous buttons don't have any functionality
<img width="655" alt="Screenshot 2023-07-05 at 11 15 29 AM" src="https://github.com/ploomber/jupysql-plugin/assets/123580782/d2627bd6-44ae-4615-95ac-e53a91b9307f">

tonykploomber commented:
Or we can just keep cancel button 
<img width="658" alt="Screenshot 2023-07-05 at 11 36 40 AM" src="https://github.com/ploomber/jupysql-plugin/assets/123580782/2cd36251-4749-4b9a-be65-dc6a88b7b2e1">

edublancas commented on environment.yml:
```diff
@@ -14,3 +14,4 @@ dependencies:
     - twine
     - hatch
     - jupysql
+    - ploomber-core
```
this is for the development environment. you need to edit: https://github.com/ploomber/jupysql-plugin/blob/2549029aad8e887c93c24926e8a21855b7fbf5be/pyproject.toml#L25

edublancas commented on jupysql_plugin/server_handlers/dashboard.py:
```diff
@@ -0,0 +1,96 @@
+BACKEND_ENDPOINT = "https://cloudapi.ploomber.io"
+
+
+class RouteHandler(APIHandler):
```
RouteHandler is pretty generic, please add a more descriptive name
